trestle/core/catalog_interface.py

```diff
@@ -362,38 +396,84 @@ def _update_all_controls_in_list(self, controls: List[cat.Control]) -> List[cat.
-                new_groups.append(sub_group)
-            group.groups = new_groups
+        group.controls = self._update_all_controls_in_list(as_list(group.controls))
+        group.controls = none_if_empty(group.controls)
+        new_groups: List[cat.Group] = []
+        for sub_group in as_list(group.groups):
+            self._update_all_controls_in_group(sub_group)
+            new_groups.append(sub_group)
+        group.groups = none_if_empty(new_groups)
+
+    def _insert_control_in_catalog(self, control_handle: ControlHandle) -> None:
+        """Insert the control into the catalog based on its path."""
+        path = control_handle.path
+        node = self._catalog
+        if path[0] != '':
+            for group_id in path:
+                found_group = None
+                for group in as_list(node.groups):
+                    if group.id == group_id:
+                        found_group = group
+                        break
+                if found_group:
+                    node = found_group
+                else:
+                    raise TrestleError(f'No controls found in catalog for group {group.id}')
+            node.title = control_handle.group_title
+            node.class_ = control_handle.group_class
+        node.controls = as_list(node.controls)
+        try:
+            index = [control.id for control in as_list(node.controls)].index(control_handle.control.id)
+            del node.controls[index]
+        except ValueError:
+            pass
+        node.controls.append(control_handle.control)
+        node.controls = none_if_empty(sorted(node.controls, key=lambda control: ControlIOWriter.get_sort_id(control)))
 
     def update_catalog_controls(self) -> None:
-        """Update the actual catalog by pulling fresh controls from the dict."""
-        if self._catalog.groups:
-            for group in self._catalog.groups:
-                self._update_all_controls_in_group(group)
-        if self._catalog.controls:
-            self._catalog.controls = self._update_all_controls_in_list(self._catalog.controls)
+        """
+        Update the actual catalog by pulling fresh controls from the dict.
+
+        During assembly, controls may be added, but not children of controls.
+        New groups may not be added.
+        A control containing controls cannot be added.  Controls containing controls are only available if
+        the parent catalog was loaded from json.
+        """
+        # first go through the catalog and pull existing controls from the dict
+        for group in as_list(self._catalog.groups):
+            self._update_all_controls_in_group(group)
+
+        new_list = []
+        for control in as_list(self._catalog.controls):
+            new_control = self.get_control(control.id)
+            new_control.controls = self._update_all_controls_in_list(as_list(control.controls))
+            new_control.controls = none_if_empty(new_control.controls)
```

Comment:
I wonder if that is something that can be done via pydantic during field initialisation. Potentially via [validators](https://pydantic-docs.helpmanual.io/usage/validators/) That would add consistency to the code, clean up all `none_if_empty` in the code, but will require to rewrite the generated models and add extra code there.

Not necessary change for this PR, but would be nice to have I think?

Reply:
I'm not sure how that would work but happy to talk about it.

Reply:
I was hoping something like this would work (inside of the `Catalog`):
```
@validator("controls")
    def non_empty_controls(cls, values):
        values = values if values else None
        return values
```

But after some experimentation it seems like reassignment like `catalog.controls = new_controls` does not trigger the validator again. So this solution will not work. There could be some elegant solution in Pydantic for this, but it is fine for now